python/lsst/afw/table/testUtils.py

```diff
@@ -0,0 +1,88 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+from __future__ import absolute_import, division, print_function
+
+import lsst.utils.tests
+import difflib
+from .tableLib import Schema
+
+__all__ = ["assertSchemasEqual", "diffSchemas"]
+
+def diffSchemas(schema1, schema2, flags=Schema.IDENTICAL):
+    """Return a string diff of two schemas.
+
+    Parameters
+    ----------
+    schema1 : :py:class:`lsst.afw.table.Schema`
+        First schema to diff.  Items appearing only in this schema
+        will be prefixed with "-" in the diff.
+    schema2 : :py:class:`lsst.afw.table.Schema`
+        Second schema to diff.  Items appearing only in this schema
+        will be prefixed with "-" in the diff.
+    flags : `int`
+        A bitwise OR of :py:class:`lsst.afw.table.Schema.ComparisonFlags`
+        indicating which features of schema items to compare.  The returned
+        diff will always show all differences, but no diff will be shown if
+        the only differences are not included in the flags.  Default is
+        `lsst.afw.table.Schema.IDENTICAL`, which checks everything.
+
+    Returns
+    -------
+    diff : `str`
+        A "unified diff" string representation of the difference between the
+        schemas, or an empty string if there is no difference.
+    """
+    result = schema1.compare(schema2, flags)
+    if result == flags:
+        return ""
+    components = []
+    if not result & Schema.EQUAL_KEYS:
+        components.append("keys")
+    if not result & Schema.EQUAL_NAMES:
+        components.append("names")
+    if not result & Schema.EQUAL_DOCS:
+        components.append("docs")
+    if not result & Schema.EQUAL_UNITS:
+        components.append("units")
+    if not result & Schema.EQUAL_ALIASES:
+        components.append("aliases")
+    diff = "\n".join(difflib.unified_diff(str(schema1).split("\n"), str(schema2).split("\n")))
+    if len(components) == 1:
+        msg = "%s differ:\n%s" % (components[0].capitalize(), diff)
+    elif len(components) == 2:
+        msg = "%s and %s differ:\n%s" % (components[0].capitalize(), components[1], diff)
+    else:
+        msg = "%s, %s, and %s differ:\n%s" % (components[0].capitalize(),
+                                              ", ".join(components[1:-1]),
+                                              components[-1], diff)
+    return msg
```

Comment:
I think this would be a bit cleaner if you made a separate function that created the list of components (with or without capitalization--that could always be done afterwards since it would be a short string). Whether the gain is worth the cost of further coding is a valid question.